src/faillog.c

```diff
@@ -83,42 +83,63 @@ usage (int status)
 	exit (status);
 }
 
-static void print_one (/*@null@*/const struct passwd *pw, bool force)
+/*
+ * Looks up the offset in the faillog file for the given uid.
+ * Returns -1 on error.
+ */
+static off_t lookup_faillog(struct faillog *fl, uid_t uid)
 {
-	static bool once = false;
-	struct tm *tm;
 	off_t offset;
-	struct faillog fl;
-	time_t now;
-	char *cp;
-	char ptime[80];
 
-	if (NULL == pw) {
-		return;
+	/* Ensure multiplication does not overflow and retrieving a wrong entry */
+	if (__builtin_mul_overflow(uid, sizeof(*fl), &offset)) {
+		fprintf(stderr,
+		        _("%s: Failed to get the entry for UID %lu\n"),
+		        Prog, (unsigned long int)uid);
+		return -1;
 	}
 
-	offset = (off_t) pw->pw_uid * sizeof (fl);
-	if (offset + sizeof (fl) <= statbuf.st_size) {
+	if (statbuf.st_size >= (off_t) sizeof(*fl) && offset <= statbuf.st_size - (off_t) sizeof(*fl)) {
```

Comment:
Please respect the 80-col margin.  Only trivial code should escape that limit.  I'm having problems reviewing this line of code.

```c
	if (statbuf.st_size >= (off_t) sizeof(*fl) &&
	    offset <= statbuf.st_size - (off_t) sizeof(*fl))
	{
```

That would fit my terminal nicely :)

Reply:
In any case, you could use `__builtin_add_overflow(offset, sizeof(*fl), &(off_t){})` here?  (I didn't test it, though; be cautious).

Reply:
@ikerexxe the old comment is outdated.  This is the only one remaining from my side.

Reply:
applied your suggestions